src/video/wayland/SDL_waylandclipboard.c

```diff
@@ -22,10 +22,66 @@
 
 #ifdef SDL_VIDEO_DRIVER_WAYLAND
 
+#include "../../events/SDL_events_c.h"
 #include "SDL_waylanddatamanager.h"
 #include "SDL_waylandevents_c.h"
 #include "SDL_waylandclipboard.h"
 
+int Wayland_SetClipboardData(_THIS, SDL_ClipboardDataCallback callback, size_t mime_count, const char **mime_types,
+                             void *userdata)
+{
+    SDL_VideoData *video_data = NULL;
+    SDL_WaylandDataDevice *data_device = NULL;
+
+    int status = 0;
+
+    if (_this == NULL || _this->driverdata == NULL) {
+        status = SDL_SetError("Video driver uninitialized");
+    } else {
+        video_data = _this->driverdata;
+        if (video_data->input != NULL && video_data->input->data_device != NULL) {
+            data_device = video_data->input->data_device;
+
+            if (callback && mime_types) {
+                SDL_WaylandDataSource *source = Wayland_data_source_create(_this);
+                Wayland_data_source_set_callback(source, callback, userdata, SDL_FALSE);
+
+                status = Wayland_data_device_set_selection(data_device, source, mime_count, mime_types);
+                if (status != 0) {
+                    Wayland_data_source_destroy(source);
+                }
+            } else {
+                status = Wayland_data_device_clear_selection(data_device);
+            }
+        }
+    }
+
+    return status;
+}
+
+#define TEXT_MIME_TYPES_LEN 5
+static const char *text_mime_types[TEXT_MIME_TYPES_LEN] = {
+    TEXT_MIME,
+    "text/plain",
+    "TEXT",
+    "UTF8_STRING",
+    "STRING",
+};
+
+static void *Wayland_ClipboardTextCallback(size_t *length, const char *mime_type, void *userdata)
+{
+    void *data = NULL;
+    *length = 0;
+
+    if (userdata != NULL) {
+        char *text = userdata;
+        *length = SDL_strlen(text);
+        data = userdata;
+    }
+
+    return data;
+}
```

Comment:
I wonder if it would be a good idea to check the mime-type anyways. Can applications request unadvertised mime-types?

Reply:
I can't find any definite answer in the docs if Wayland blocks this or we have to deal with it so we might as well check on our side to be safe.